packages/messaging/src/helpers/registerDefaultSw.ts

```diff
@@ -39,9 +39,40 @@ export async function registerDefaultSw(
     messaging.swRegistration.update().catch(() => {
       /* it is non blocking and we don't care if it failed */
     });
+    await waitForRegistrationActive(messaging.swRegistration);
   } catch (e) {
     throw ERROR_FACTORY.create(ErrorCode.FAILED_DEFAULT_REGISTRATION, {
       browserErrorMessage: (e as Error)?.message
     });
   }
 }
+
+/**
+ * Waits for registration to become active. MDN documentation claims that
+ * a service worker registration should be ready to use after awaiting
+ * navigator.serviceWorker.register() but that doesn't seem to be the case in
+ * practice, causing the SDK to throw errors when calling
+ * swRegistration.pushManager.subscribe() too soon after register(). The only
+ * solution seems to be waiting for the service worker registration `state`
+ * to become "active".
+ */
+async function waitForRegistrationActive(
+  registration: ServiceWorkerRegistration
+): Promise<void> {
+  return new Promise<void>((resolve, reject) => {
+    if (registration.active) {
+      resolve();
+    }
+    const incomingSw = registration.installing || registration.waiting;
+    if (incomingSw) {
+      incomingSw.onstatechange = ev => {
+        if ((ev.target as ServiceWorker)?.state === 'activated') {
+          incomingSw.onstatechange = null;
+          resolve();
+        }
```

Comment:
What happens here if the service worker was installing but ran into an error (could be a JS error) and doesn't transition to the activated state?

I'm generally worried that the state may never become active, causing this to never resolve or reject. What do you think about adding a timeout to reject after some time?

Reply:
I added a 10s timeout. I originally set it up to be customizable in case people had very slow devices or something but it involved so many changes (getToken and deleteToken both need a new param) I didn't think it was a good idea for an edge case. In my testing, I had to set the timeout all the way down to 1ms in order for it to be triggered so the registration should be pretty fast?

Reply:
Sounds good!